src/Language/PureScript/Sugar/CaseDeclarations.hs

```diff
@@ -220,59 +218,57 @@ desugarGuardedExprs ss (Case scrut alternatives) =
     desugarAltOutOfLine alt_binder rem_guarded rem_alts mk_body
       | Just rem_case <- mkCaseOfRemainingGuardsAndAlts = do
 
-        desugared     <- desugarGuardedExprs ss rem_case
+        desugared     <- desugarGuardedExprs rem_case
         rem_case_id   <- freshIdent'
         unused_binder <- freshIdent'
 
         let
           goto_rem_case :: Expr
-          goto_rem_case = Var (Qualified Nothing rem_case_id)
-            `App` Literal (BooleanLiteral True)
-          alt_fail = [CaseAlternative [NullBinder] [MkUnguarded goto_rem_case]]
+          goto_rem_case = App sa (Var sa (Qualified Nothing rem_case_id)) (Literal sa (BooleanLiteral True))
+          alt_fail = [CaseAlternative (fst sa) [NullBinder (fst sa)] [mkUnguarded goto_rem_case]]
 
-        pure $ Let [
-          ValueDecl (ss, []) rem_case_id Private []
-            [MkUnguarded (Abs (VarBinder unused_binder) desugared)]
+        pure $ Let sa [
+          ValueDecl sa rem_case_id Private []
+            [mkUnguarded (Abs sa (VarBinder (fst sa) unused_binder) desugared)]
           ] (mk_body alt_fail)
 
       | otherwise
       = pure $ mk_body []
       where
         mkCaseOfRemainingGuardsAndAlts
           | not (null rem_guarded)
-          = Just $ Case scrut (CaseAlternative alt_binder rem_guarded : rem_alts)
+          = Just $ Case sa scrut (CaseAlternative (fst sa) alt_binder rem_guarded : rem_alts)
           | not (null rem_alts)
-          = Just $ Case scrut rem_alts
+          = Just $ Case sa scrut rem_alts
           | otherwise
           = Nothing
 
     scrut_nullbinder :: [Binder]
-    scrut_nullbinder = replicate (length scrut) NullBinder
+    scrut_nullbinder = replicate (length scrut) (NullBinder (fst sa))
```

Comment:
All the spans from line 185 till here are just the toplevel span, but I don't see how to properly break this down. Also doesn't help that the code is indented 4 wheres deep, which makes it very hard to tell what spans are actually in scope.